#24 categories edit fix

unam3 opened the pull request.

unam3 committed:
processing of same category_id and parent_id in updateCategory was added #24

unam3 committed:
getCategoryDescendants

run to test output:
-- Right {descendants: [10]}
curl -i -H 'Cookie: SESSION='${AUTHOR_ADMIN_SESSION} -X PATCH -d '{"category_id": 8, "name": "pluh_pattched", "parent_id": 9}' http://0.0.0.0:8081/categories

-- Right {descendants: null}
curl -i -H 'Cookie: SESSION='${AUTHOR_ADMIN_SESSION} -X PATCH -d '{"category_id": 8, "name": "pluh_pattched", "parent_id": 10}' http://0.0.0.0:8081/categories

-- Right {error: string}
curl -i -H 'Cookie: SESSION='${AUTHOR_ADMIN_SESSION} -X PATCH -d '{"category_id": 8, "name": "pluh_pattched", "parent_id": 1123}' http://0.0.0.0:8081/categories

unam3 committed:
isCategoryExist implementation WIP

unam3 committed:
isCategoryExist and getCategoryDescendants were chained

unam3 committed:
isCategoryExist was rewritten

unam3 committed:
getCategoryDescendants was used in updateCategory

unam3 committed:
eNoSuchCategory -> makeNoSuchCategory

unam3 committed:
eParentCategoryDoesNotExist was removed

unam3 committed:
ExceptT'zation WIP

unam3 committed:
updateCategory refactoring was done

unam3 committed:
error reporting was fixed

updateCategory tests "returns error if parent is set to itself" and "returns error if parent is set to category descendant" now can be passed.

unam3 committed:
isParentCategoryExist was updated

unam3 committed:
updateCategory checks were renamed

unam3 committed:
parentIdDescendantCheck was updated

unam3 committed:
category errors were moved to a proper place

antonkalinin-ml commented on src/RestNews/DB/Request.hs:
```diff
@@ -173,6 +175,24 @@ createCategory =
+isCategoryExist =
+    [TH.singletonStatement|
+        with isCategory as (
+            select count(true) != 0 as exist from categories where category_id = $1 :: int4
```
Если нужно проверить, существует ли какая-нибудь строка, то лучше не использовать `count()`, а лучше `select exists (...)` или `select true from ... where ... limit 1` и потом проверить, вернулось 0 строк или 1. Из-за `count` постгресу придется довести поиск до конца, хотя достаточно найти первую строку и остановиться.

Это мелочевка, править не надо. Тем более, тут он скорее всего догадается соптимизировать.

antonkalinin-ml commented on schema.psql:
```diff
@@ -68,6 +68,23 @@ create table articles_comments (
 
 commit;
 
+CREATE OR REPLACE FUNCTION public.get_category_descendants(cat_id integer)
```
Можно чуть проще сделать:

- искать предков, а не потомков. Даже если у нас 1000 категорий, то предков много не будет - больше 10 очень вряд ли. А вот потомков может быть и 1000, если взять корневую категорию.
- можно не возвращать список, а принять две категории и определить, является ли одна предком второй. Тогда запрос будет побыстрее и не придется гонять список категорий из базы в приложение.

Но так тоже хорошо, пусть остается.

antonkalinin-ml commented on src/RestNews/DB/ProcessRequest.hs:
```diff
@@ -212,32 +214,131 @@ createCategory sessionRun connection createCategoryRequest = do
+        Right isCategoryExist' ->
+            if isCategoryExist'
+                then Right ()
+                else Left . Right . encode . makeNoSuchCategory . pack $ show parent_id'
```
```suggestion
        Right True -> Right ()
        Right False -> Left . Right . encode . makeNoSuchCategory . pack $ show parent_id'
```

antonkalinin-ml commented on src/RestNews/DB/ProcessRequest.hs:
```diff
@@ -212,32 +214,131 @@ createCategory sessionRun connection createCategoryRequest = do
+    Connection
+    -> Int32
+    -> m (Either Session.QueryError Bool)
+isCategoryExist connection category_id' =
```
С первого прочтения трудно понять, чем отличаются имена `isCategoryExist` и `parentCategoryExistCheck` - вроде оба про одно и то же. Первая функция нужна внутри второй, а сама по себе она больше нигде не используется. Дабы читатель не раздумывал, которую из них положено вызывать, можно лишить самостоятельности `isCategoryExist`, переместив ее в блок `where` функции `parentCategoryExistCheck`.

antonkalinin-ml commented on src/RestNews/DB/ProcessRequest.hs:
```diff
@@ -212,32 +214,131 @@ createCategory sessionRun connection createCategoryRequest = do
+
+        Left sessionError -> Left . Left $ show sessionError
+
+        Right descendants ->
+            let parentIdDescendantCheck' = Data.Vector.elem parent_id' descendants
+            in if parentIdDescendantCheck'
+                then Left . Right $ encode eParentIdIsDescendant
+                else Right ()
```
Внутри `case` можно не писать `if`, кейс умеет булевы условия:

```suggestion
        Right descendants
            | parent_id' `Data.Vector.elem` descendants -> Left . Right $ encode eParentIdIsDescendant
            | otherwise -> Right ()
```

И еще `elem` часто пишут инфиксно, но это кому как нравится.

antonkalinin-ml commented on src/RestNews/DB/ProcessRequest.hs:
```diff
@@ -212,32 +214,131 @@ createCategory sessionRun connection createCategoryRequest = do
                 _ -> H . Left $ show sessionError
         )
 
-updateCategory :: MonadIO m =>
+
+updateCategory' :: MonadIO m =>
     (Session.Session Value -> Connection -> m (Either Session.QueryError Value))
     -> Connection
-    -> UpdateCategoryRequest
+    -> (Int32, Text, Maybe Int32)
```
Жаль, что больше не используется рекорд `UpdateCategoryRequest`. С ним понятнее, чем с кортежем из загадочных строк и интов, где надо сматчиться в правильном порядке и ничего не перепутать.

Возможно, тебе надоело возиться с рекордами, везде указывать полный тип, например `category_id (updateCategoryRequest :: UpdateCategoryRequest)`. Есть такая проблема, решается префиксами/квал. импортами (костыль, но с этим много где живут, и `aeson` можно уговорить отрезать префиксы) или `RecordWildCards` + `NamedFieldPuns`. Примерно так:

```hs
{-# LANGUAGE RecordWildCards #-}
...
updateCategory sessionRun connection UpdateCategoryRequest {..} = do
  -- внутри можно использовать имена полей UpdateCategoryRequest как переменные
```

или

```hs
{-# LANGUAGE NamedFieldPuns #-}
...
updateCategory sessionRun connection UpdateCategoryRequest{category_id, name} = do
  -- category_id, name доступны как переменная
```

Есть способы менее костыльные, но сложнее. В ghc 9.2 эту проблему решили, теперь надо дожить до стабильного релиза.

antonkalinin-ml commented:
Спасибо, все проверки добавлены, круто :)

unam3 committed:
if-conditionals inside case were refactored

https://github.com/unam3/fsd-rest-news/pull/33#discussion_r792068431
https://github.com/unam3/fsd-rest-news/pull/33#discussion_r792081527

unam3 committed:
isCategoryExist definition was moved into parentCategoryExistCheck

https://github.com/unam3/fsd-rest-news/pull/33#discussion_r792078382

unam3 committed:
brings UpdateCategoryRequest back

https://github.com/unam3/fsd-rest-news/pull/33#discussion_r792086662

antonkalinin-ml commented on src/RestNews/DB/ProcessRequest.hs:
```diff
@@ -254,23 +254,20 @@ sameCategoryIdCheck category_id' parent_id' =
 parentCategoryExistCheck connection parent_id' = do
-    
-    eitherIsCategoryExist <- isCategoryExist connection parent_id'
+    let {
```
А зачем фигурные скобки? Кажется, без них скомпилилось бы не хуже. Отступы вроде правильно стоят.

unam3 replied:
Скорее всего вначале отступы неверно стояли и в ходе экспериментов скомпилилось с фигурными.

antonkalinin-ml reviewed (approved):
Спасибо за правки, у меня больше вопросов нет, можно мержить.

unam3 committed:
redundant brackets were removed

https://github.com/unam3/fsd-rest-news/pull/33#pullrequestreview-870339420
